scene/gui/rich_text_label.cpp

```diff
@@ -4459,6 +4460,23 @@ int RichTextLabel::get_visible_characters() const {
 	return visible_characters;
 }
 
+int RichTextLabel::get_character_line(int p_char) {
+	int line_count = 0;
+	if (0 < p_char && p_char <= get_total_character_count()) {
+		int char_count = 0;
+		for (int i = 0; i < main->lines.size(); i++) {
+			for (int j = 0; j < main->lines[i].text_buf->get_line_count(); j++) {
+				line_count++;
+				char_count += TS->shaped_text_get_glyph_count(main->lines[i].text_buf->get_line_rid(j));
```

Comment:
No, this is not a character count, but a glyph count, e.g., depending on font `Ē` might be two glyphs `E` and macron, and `ffi` might be one ligature glyph.

`main->lines[i]` have a `char_offset` and `char_count` fields you can use.
You can get a range of characters on each auto-wrapped line (relative to the line) using:
```c++
Vector2i range = TS->font_render_range(main->lines[i].text_buf->get_line_rid(j));
```

Reply:
Ah. That makes sense.
I was originally exploring `char_offset` and `char_count`, but they provide the count of characters of all lines in a `TextParagraph` and not per line, so they unfortunately couldn't be used.
Did you mean `shaped_text_get_range()`? `font_render_range()` takes four parameters.

Reply:
> Did you mean shaped_text_get_range()? font_render_range() takes four parameters.

Yes, copy-pasted a wrong line.

Something like this should work (counting lines and characters from `0`, and returning `-1` is out of range):

```c++
int RichTextLabel::get_character_line(int p_char) {
	int line_count = 0;
	for (int i = 0; i < main->lines.size(); i++) {
		if (p_char >= main->lines[i].char_offset && p_char < main->lines[i].char_offset + main->lines[i].char_count) {
			for (int j = 0; j < main->lines[i].text_buf->get_line_count(); j++) {
				Vector2i range = TS->shaped_text_get_range(main->lines[i].text_buf->get_line_rid(j));
				if (p_char >= main->lines[i].char_offset + range.x && p_char < main->lines[i].char_offset + range.y) {
					return line_count;
				}
				line_count++;
			}
		} else {
			line_count += main->lines[i].text_buf->get_line_count();
		}
	}
	return -1;
}
```

And for consistency with other `line` function, it might be useful to also add a second function that is not counting auto-wraps:

```c++

int RichTextLabel::get_character_paragraph(int p_char) {
	int para_count = 0;
	for (int i = 0; i < main->lines.size(); i++) {
		if (p_char >= main->lines[i].char_offset && p_char < main->lines[i].char_offset + main->lines[i].char_count) {
			return para_count;
		} else {
			para_count++;
		}
	}
	return -1;
}
```

Reply:
You're right. The supplemental method would be helpful, since otherwise `get_character_line()` only returns the paragraph number if auto-wrap is off.

I was under the assumption line numbering wasn't 0 based. I appreciate that you corrected the `line_count` increment to after the early return.

Reply:
The only change I needed to make that were the if statements and I'm now getting consistent results with all `visible_character_behavior` settings.

```csharp
int RichTextLabel::get_character_line(int p_char) {
	int line_count = 0;
	for (int i = 0; i < main->lines.size(); i++) {
		if (main->lines[i].char_offset < p_char && p_char <= main->lines[i].char_offset + main->lines[i].char_count) {
			for (int j = 0; j < main->lines[i].text_buf->get_line_count(); j++) {
				Vector2i range = TS->shaped_text_get_range(main->lines[i].text_buf->get_line_rid(j));
				if (main->lines[i].char_offset + range.x < p_char && p_char <= main->lines[i].char_offset + range.y) {
					return line_count;
				}
				line_count++;
			}
		} else {
			line_count += main->lines[i].text_buf->get_line_count();
		}
	}
	return -1;
}
```